internal/RF_Base.ecl

```diff
@@ -0,0 +1,553 @@
+  EXPORT DATASET(TreeNodeDat) GetLeafsForData(DATASET(TreeNodeDat) tNodes, DATASET(GenField) X) := FUNCTION
+    // Distribute X by wi and id.
+    xD := DISTRIBUTE(X, HASH32(wi, id));
+    // Extend each root for each ID in X
+    // Leave the extended roots distributed by wi, id.
+    roots := tNodes(level = 1);
+    rootsExt := JOIN(xD(number=1), roots, LEFT.wi = RIGHT.wi, TRANSFORM(TreeNodeDat, SELF.id := LEFT.id, SELF := RIGHT),
+                     MANY, LOOKUP);
+    rootBranches := rootsExt(number != 0); // Roots are almost always branch (split) nodes.
+    rootLeafs := rootsExt(number = 0); // Unusual but not impossible
+    loopBody(DATASET(TreeNodeDat) levelBranches, UNSIGNED tLevel) := FUNCTION
+      // At this point, we have one record per node, per id.
+      // We extend each id down the tree one level at a time, picking the correct next nodes
+      // for that id at each branch.
+      // Next nodes are returned -- both leafs and branches.  The leafs are filtered out by the LOOP,
+      // while the branches are send on to the next round.
+      // Ultimately, a leaf is returned for each id, which defines our final result.
+      // Select the next nodes by combining the selected data field with each node
+      // Note that:  1) we retain the id from the previous round, but the field number(number) is derived from the branch
+      //             2) 'value' in the node is the value to split upon, while value in the data (X) is the value of that datapoint
+      //             3) NodeIds at level n + 1 are deterministic.  The child nodes at the next level's nodeId is 2 * nodeId -1 for the
+      //                left node, and 2 * nodeId for the right node.
+      branchVals := JOIN(levelBranches, xD, LEFT.wi = RIGHT.wi AND LEFT.id = RIGHT.id AND LEFT.number = RIGHT.number,
+                          TRANSFORM({TreeNodeDat, BOOLEAN branchLeft},
+                                      SELF.branchLeft :=  ((LEFT.isOrdinal AND RIGHT.value <= LEFT.value) OR
+                                                          ((NOT LEFT.isOrdinal) AND RIGHT.value = LEFT.value)),
+                                      SELF.parentId := LEFT.nodeId, SELF := LEFT),
+                          LOCAL);
+      // Now, nextNode indicates the selected (left or right) nodeId at the next level for each branch
+      // Now we use nextNode to select the node for the next round, for each instance
+      nextLevelNodes := tNodes(level = tLevel + 1);
+      nextLevelSelNodes := JOIN(branchVals, nextLevelNodes, LEFT.wi = RIGHT.wi AND
+                              LEFT.treeId = RIGHT.treeId AND LEFT.nodeId = RIGHT.parentId AND
+                              LEFT.branchLeft = RIGHT.isLeft,
+                              TRANSFORM(TreeNodeDat, SELF.id := LEFT.id, SELF := RIGHT), LOOKUP);
+      // Return the selected nodes at the next level.  These nodes may be leafs or branches.
```

Comment:
Anyway to have tNodes so that the records selected don't need to be replicated and distributed inside of the loop?  Not much of a speed bump though as the max depth is only double digits.